File: python/logger/log_types.py
import os
import socket
import sys
import time
from typing import Union
from enum import Enum

class Level(Enum):
    """
    Levels Enum for different log levels
    """
    INFO = 0
    DEBUG = 1
    WARNING = 2
    ERROR = 3
    CRITICAL = 4


class Log:
    """
    Base class for log types.
    """
    def get_message(self) -> str:
        raise NotImplementedError("Log class must implement the get_message method.")

    def get_timestamp(self) -> str:
        raise NotImplementedError("Log class must implement the get_timestamp method.")

class MessageLog(Log):
    """
    Log class for messages.
    """
    def __init__(self, app_name: str, timestamp: str, level: Level, message: str) -> None:
        self.app_name = app_name
        self.timestamp = timestamp
        self.level = level
        if not isinstance(message, str):
            raise ValueError("MessageLog 'message' must be of type str.")
        self.message = message
    
    def get_message(self) -> str:
        return f"[{self.timestamp}] [{self.app_name}] [{self.level.name}] {self.message}"
    
    def get_timestamp(self) -> str:
        return self.timestamp

class DataLog(Log):
    """
    Log class for data.
    """
    def __init__(self, app_name: str, timestamp: str, name: str, value: Union[float, int]) -> None:
        self.app_name = app_name
        self.timestamp = timestamp
        if not isinstance(name, str):
            raise ValueError("DataLog 'name' must be of type str.")
        self.name = name
        if not isinstance(value, (float, int)):
            raise ValueError("DataLog 'value' must be of type float or int.")
        self.value = value
    
    def get_message(self) -> str:
        return f"[{self.timestamp}] [{self.app_name}] [{self.name}] {self.value}"
    
    def get_timestamp(self) -> str:
        return self.timestamp
    
class Handler:
    """
    Base Handler class.

    Defines the interface for all log handlers.
    """
    def send(self, message: str):
        raise NotImplementedError("Log handler must implement the send method.")


class ConsoleHandler(Handler):
    """
    Console Handler class for logging.

    Prints logs to console.
    """
    def send(self, log: Log):
        print(log.get_message())

class ServerHandler(Handler):
    """
    Server Handler class for logging.

    Sends logs to centralised server. This is done through http requests.
    """
    def __init__(self) -> None:
        super().__init__()
        self.node_id = None

    def setup(self, app_name: str) -> str:
        """
        Setup the logger with the given app name.

        Args:
            app_name (str): Name of the application.

        Returns:
            node_id (str): Unique node id for this logger.
        """
        ip = socket.gethostbyname(socket.gethostname())
        time_now = time.time()
        self.node_id = hash(f"{app_name}_{ip}_{time_now}") % ((sys.maxsize + 1) * 2)
        self.log_dir = "/tmp/Logger"
        self.file_name = f"{self.node_id}.log"
        os.makedirs(self.log_dir, exist_ok=True)
        print("Temporary log file path:", self.log_dir)

    def send(self, log: Log):
        if not self.node_id or not self.log_dir:
            raise Exception("Logger not setup. Call setup() before sending logs.")

        log_type = log.__class__.__name__
        with open(self.log_dir + "/" + self.file_name, "a") as f:
            f.write(f"[{self.node_id}] [{log_type}] {log.get_message()}\n")

File: go/logger/log_types.go
// logger/log_types.go

package logger

import (
	"fmt"
)

// Level represents the logging level
type Level int

const (
	INFO Level = iota
	DEBUG
	WARNING
	ERROR
	CRITICAL
)

// Log interface to define log structure
type Log interface {
	GetMessage() string
	GetTimestamp() string
}

// MessageLog represents a log message
type MessageLog struct {
	AppName   string
	Timestamp string
	Level     Level
	Message   string
}

func (m *MessageLog) GetMessage() string {
	return fmt.Sprintf("[%s] [%s] [%s] %s", m.Timestamp, m.AppName, m.Level.String(), m.Message)
}

func (m *MessageLog) GetTimestamp() string {
	return m.Timestamp
}

// DataLog represents a data log
type DataLog struct {
	AppName   string
	Timestamp string
	Name      string
	Value     float64
}

func (d *DataLog) GetMessage() string {
	return fmt.Sprintf("[%s] [%s] [%s] %v", d.Timestamp, d.AppName, d.Name, d.Value)
}

func (d *DataLog) GetTimestamp() string {
	return d.Timestamp
}

// String representation for the Level type
func (l Level) String() string {
	switch l {
	case INFO:
		return "INFO"
	case DEBUG:
		return "DEBUG"
	case WARNING:
		return "WARNING"
	case ERROR:
		return "ERROR"
	case CRITICAL:
		return "CRITICAL"
	default:
		return "UNKNOWN"
	}
}